### PR TITLE
Added the ability to define tags for elasticache

### LINKED ISSUE #24214: Want to set tags on elasticache clusters
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
elasticache

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /home/lyle/projects/prima_cm/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.9 (default, Jun 29 2016, 13:08:31) [GCC 4.9.2]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
If I want to have duplicate environments for staging and production, it's easy to distinguish them when grouping by tags. I also want to provision AWS with Ansible, but the elasticache module does not have a `tags` parameter.

### PR DESCRIPTION
**Pull request by memo42**
##### SUMMARY

Added the ability to add tags. Fixes #24214

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

elasticache

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 3142e60693) last updated 2017/08/09 11:02:06 (GMT +200)
  config file = /home/mo/playbooks/production.cfg
  configured module search path = [u'/home/mo/playbooks/library']
  ansible python module location = /home/mo/ansible/lib/ansible
  executable location = /home/mo/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 1]
```diff
@@ -128,6 +135,25 @@
     name: "test-please-delete"
     state: rebooted
 
+# Create cache cluster with Tags
+- elasticache:
+    name: "test-please-delete"
+    state: present
+    engine: redis
+    tags: [
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py]: This should use standard ansible tag form

```
tags:
  test: testa
  testb: testc
  testd: testf
```

You'll want to use `ansible_dict_to_boto3_tag_list` for the conversion

**Review by willthames — CHANGES_REQUESTED** [review]
There is a lot of work involved in managing tags. Have a look at e.g. ec2_group.

In particular:
* Adding tags on new resources
* Updating tags on existing resources
* Tracking `changed` if a tag gets added or amended
* Using `purge_tags` to remove tags only if set

Have a look at `compare_aws_tags` too to manage tag amendment.

I also strongly recommend adding test cases around this too. Again, ec2_group is a good model.

File: lib/ansible/modules/cloud/amazon/elasticache.py (hunk 1 above)
Comment (memo42) [reply lib/ansible/modules/cloud/amazon/elasticache.py]: OK, thanks

**Comment by memo42** [comment]
So your requests should be met now. I tried getting started with tests, but ran into bigger problems ( Issue #28351 ), so that might take a litte bit longer.

**Comment by willthames** [comment]
Sorry, I've only just noticed this again (was looking for some other elasticache features). I'll endeavour to test this out shortly

**Comment by memo42** [comment]
Nice, thanks. I did not have time to properly get into your test-setup, yet. I did get the local tests running, but there was no security group template for elasticache, my first, hacky attempt of building one did not work and I was not able to try again since.

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 2]
```diff
@@ -182,6 +210,37 @@ def ensure_present(self):
...
+    def get_arn(self):
+        region, ec2_url, aws_connect_params = get_aws_connection_info(self.module, boto3=True)
+        if region:
+            client = boto3_conn(self.module, conn_type='client', resource='rds',
+                                region=region, endpoint=ec2_url, **aws_connect_params)
+        else:
+            self.module.fail_json(msg="region must be specified")
+        instance_counts = {}
+        response = client.describe_db_instances()
+        user_id = response['DBInstances'][0]['DBInstanceArn'].split(':')[4]
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py]: Argh, it's so frustrating that the Elasticache API doesn't return ARNs!

This is probably the best way to get user ID though:
https://stackoverflow.com/a/10198870/3538079

There is always a chance that a user allowed to use Elasticache isn't allowed to use `rds.describe_db_instances`, so it's best to rely on as few other APIs as possible (I think everyone has permission to run `iam.get_caller_identity`)

**Comment by willthames** [comment]
Please can you commit your tests too (even if they're failing, I can always look at improving them)

**Comment by willthames** [comment]
Also, if you do have problems setting up tests that my blog post doesn't help with, let me know so that I can update my blog post! (Based on #28351 I should document that awscli is required!)

File: lib/ansible/modules/cloud/amazon/elasticache.py (hunk 2 above)
Comment (memo42) [reply lib/ansible/modules/cloud/amazon/elasticache.py]: cool, thanks

**Comment by memo42** [comment]
Thanks for your help.
I fixed the getARN method the way you suggested.
I merged in the current devel branch, which ansibot did not like so I rebased and force pushed. That is why this PR now has some weird tags. Sorry about that.
I also pushed my tests. I did not get very far, simply copied the tests from some other module, removed everything that was irrelevant and adjusted a few details.
The problem was that the playbook that created the iam policy did not have a template for elasticache permissions. I tried adding one that allows all for a first test, but it did not work.
I probably just need to spend some time with this to figure everything out.
Not sure when I will get to that, though.

**Comment by yuwzho** [comment]
-label azure

**Comment by dharmabumstead** [comment]
Closing this one. Feel free to rebase and reopen.

File: hacking/aws_config/testing_policies/elasticache-policy.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{# Note that not all EC2 API Actions allow a specific resource #}
```
Comment (willthames) [inline hacking/aws_config/testing_policies/elasticache-policy.json:1]: Please don't add a new policy - reuse storage or compute for elasticache (we can only attach 10 policies to a group, so we need to limit the number we have)

File: hacking/aws_config/testing_policies/elasticache-policy.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+{# See http://docs.aws.amazon.com/AWSEC2/latest/APIReference/ec2-api-permissions.html#ec2-api-unsupported-resource-permissions #}
```
Comment (willthames) [inline hacking/aws_config/testing_policies/elasticache-policy.json:2]: this comment is unrelated to elasticache

File: hacking/aws_config/testing_policies/elasticache-policy.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "AllowUnspecifiedEC2Resource",
```
Comment (willthames) [inline hacking/aws_config/testing_policies/elasticache-policy.json:7]: Please update the SID for Elasticache

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 6]
```diff
@@ -95,6 +95,21 @@
     required: false
     default: no
     choices: [ "yes", "no" ]
+  tags:
+    description:
+      - Set instance tags
+    required: false
+    default: none
+    version_added: "2.5"
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py:103]: needs to be 2.6 now

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 7]
```diff
@@ -182,6 +210,35 @@ def ensure_present(self):
...
+        else:
+            self.module.fail_json(msg="region must be specified")
+        instance_counts = {}
+        response = client.get_caller_identity()
+        user_id = response['Account']
+        arn = "arn:aws:elasticache:" + region + ":" + user_id + ":cluster:" + self.name
+        return arn
+
+    def compare_and_update_tags(self):
+        if self.tags:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py:227]: This check shouldn't be required if `self.tags` defaults to the empty dict - the else clause would be taken care of in the same way using compare_aws_tags.

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 8]
```diff
@@ -386,7 +447,7 @@ def _requires_modification(self):
             'EngineVersion': self.cache_engine_version
         }
         for key, value in modifiable_data.items():
-            if value is not None and self.data[key] != value:
+            if value and self.data[key] != value:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py:450]: the change from `value is not None` is very dangerous if any value can be `False` instead.

File: lib/ansible/modules/cloud/amazon/elasticache.py [hunk 9]
```diff
@@ -420,7 +481,7 @@ def _requires_destroy_and_create(self):
         if self.zone is not None:
             unmodifiable_data['zone'] = self.data['PreferredAvailabilityZone']
         for key, value in unmodifiable_data.items():
-            if getattr(self, key) is not None and getattr(self, key) != value:
+            if getattr(self, key) and getattr(self, key) != value:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache.py:484]: Again, this can be dangerous if any value can be `False`

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by memo42** [comment]
I no longer have access to the repository and will create a new pull request
